Include json formatter

fede-moya opened the pull request:
As discussed in [https://github.com/codeclimate/test-reporter/issues/413](https://github.com/codeclimate/test-reporter/issues/413)  , this PR has the intention of both adding the capability to modify the formatter through an environmental variable and also to include a json formatter by default.

We are suggesting to use the following json formatter https://github.com/codeclimate-community/simplecov_json_formatter. 

Additionally, as part of the original discussion, it would be the extreme help for those developers using CodeClimate services if we add the possibility to output JSON when the `CC_TEST_REPORTER_ID` is present. 

---
**TO-DO**

- [x] Get agreement on the implementation
- [x] Add unit tests
- [x] Add cucumber/feature tests
- [x] Add documentation to express this new functionality

fede-moya commented:
@PragTob Hello Tobias, how are you ? What do you think of the implementation ?

PragTob commented:
:wave: 

Hi there, sorry I'm not getting much open source work in these days during the week, I'll hopefully take a look on Saturday. Sorry!

Tobi

PragTob reviewed (commented):
Hi, thanks for the PR :tada: 

See the inline comments, but most of them are just about how the goal should be accomplished and that I'd rather keep the environment variable magic out of there for now :)

also sorry for the CI, I've got no idea what's wrong with it.

The more complicated matter is the question of where the code should live. I think we might have had a misunderstanding over at the test-reporter issue:

I want to move this to a mono repo with `simplecov` and `simplecov-html` and potentially `simplecov-core` or something like that. The problem I've realized with having the repos separate is that they're usually so tightly integrated that they break all too easy with unintended changes (especially in core). 

With HTML it's even a bit worse since its tests live here, but ultimately the JSON acceptance/integration tests would probably also live here because the setup/helpers for it are already here.

From a maintenance/me work perspecitve I'd rather have the repo with you and me not worrying about it of course, but from the perspective of this shipping as a default functionality of SimpleCov now that might break things I need at least push access (git and gem) (not just me but other SimpleCov maintainers as well).

Still, there's always worry about it breaking. In my experience the easiest way to achieve this is have it in a mono repo were all tests run once anything has changed. Of course I haven't done any mono repo migration yet (and I'm unsure when I'd get to it) so that'sa bit up in the air sadly.

That of course would also mean me reviewing all of the formatter code (as it'd land in her) and having some tests for it in our cucumber/aruba setup.

What do you think?


![IMG_20190915_142942](https://user-images.githubusercontent.com/606517/92325252-5c4c3b00-f049-11ea-91ad-16b5c21e2546.jpg)

PragTob commented:
For ref about the CI the issue is t #925

fede-moya commented on doc/alternate-formatters.md:
```diff
@@ -35,8 +35,8 @@ CSV formatter for SimpleCov
-#### [simplecov-json](https://github.com/vicentllongo/simplecov-json)
-*by Vicent Llongo*
+#### [simplecov-json](https://github.com/codeclimate-community/simplecov_json_formatter)
+*by CodeClimate Community"
```
ooops, sorry I thought it was currently unmaintained 🙏🏼

fede-moya commented on lib/simplecov/default_formatter.rb:
```diff
@@ -0,0 +1,48 @@
+
+DEFAULT_FORMATTER = "HTML"
+
+def constantize_formatter(formatter)
```
Makes sense 👍🏼

fede-moya commented on simplecov.gemspec:
```diff
@@ -37,6 +37,7 @@ Gem::Specification.new do |gem|
 
   gem.add_dependency "docile", "~> 1.1"
   gem.add_dependency "simplecov-html", "~> 0.11"
+  gem.add_dependency "simplecov_json_formatter", "~> 0.1.2"
```
Makes sense 👍🏼

fede-moya commented:
Thanks for taking some weekend time for this Tobias 👏🏻

Regarding the environment variable magic, makes sense to exist on it's own PR. I would also enjoy removing the use of `const_get`. So, following your feedback I extract that from this PR.

fede-moya commented:
Thanks for taking some weekend time for this Tobias 👏🏻

Regarding the environment variable magic, makes sense to exist on it's own PR. I would also enjoy removing the use of `const_get`. So, following your feedback I extract that from this PR.

fede-moya commented:
Thanks for taking some weekend time for this Tobias 👏🏻

Regarding the environment variable magic, makes sense to exist on it's own PR. I would also enjoy removing the use of `const_get`. So, following your feedback I extract that from this PR.

If one of your ideas described back in the test-reporter issue was to perhaps start the monorepo migration with having this JSON Formatter living inside this repository, sorry I might have misunderstood that. Saying that,  when reading the test reporter issue I couldn't thought on an implementation of the `simplecov_json_formatter` that could facilitate your monorepo's migration plan, though I do realize that including the `simplecov_json_formatter` will indeed increase the amount of work needed for making this repo a monorepo. We'll be happy to help you with that migration after you structure it.

fede-moya commented:
Thanks for taking some weekend time for this Tobias 👏🏻

Regarding the environment variable magic, makes sense to exist on it's own PR. I would also enjoy removing the use of `const_get`. So, following your feedback I extract that from this PR.

If one of your ideas described back in the test-reporter issue was to perhaps start the monorepo migration with having this JSON Formatter living inside this repository, sorry I might have misunderstood that. Saying that,  when reading the test reporter issue I couldn't thought on an implementation of the `simplecov_json_formatter` that could facilitate your monorepo's migration plan, though I do realize that including the `simplecov_json_formatter` will indeed increase the amount of work needed for making this repo a monorepo. We'll be happy to help you with that migration after you structure it.

I will grant you with access for both git and gem. 

Sounds good ?

fede-moya commented:
I have already added you as an owner of the gem. I think that enables you to add others.

fede-moya commented:
Feedback addressed. Removed the environment variable magic, simplified `default_formatter.rb` and load `simplecov_json_formatter` on demand.

Tobias, would you like to add a reference in the docs of this new "capabilty", to keep record of it ? I can add it, just I didn't because I wasn't sure If you want it since its not a "general feature" lets say.

fede-moya commented:
Could unit tests be enough ? or should we write some cucumber tests ?

PragTob commented:
Hi there @fede-moya ,

sorry for the silence. I've been aiming what little time I had at trying to fix the CI that magically seems to be working again (probably/maybe a new Chrome version that fixed it or they rolled something back).

Yeah having gem and repo access should be good enough for now, thank you :) 

imo it definitely belongs into the docs, at least that there is the JSON formatter now. I think that's very important to mention and good to discover. That it automatically runs for Code Climate I'd mention in its description not as a major thing.

as for tests, for very integrated features such as this I'd prefer to have cucumber/feature tests in the vain of:

* configure it to run with JSON formatter, it runs and the output is correct (we also test the output of the HTML formatter in here, which is part of the reason why that should be a mono repo :D )
* set the code climate environment variable and see that it triggers and gives the right output

PragTob commented on lib/simplecov/default_formatter.rb:
```diff
@@ -0,0 +1,20 @@
+        # When running under a CI that uses CodeClimate, JSON output is expected
+        if ENV.fetch("CC_TEST_REPORTER_ID", nil)
+          require "simplecov_json_formatter"
+          formatters.push(SimpleCov::Formatter::JSONFormatter)
```
:+1:

PragTob commented on spec/default_formatter_spec.rb:
```diff
@@ -0,0 +1,37 @@
+      end
+
+      it "returns a MultiFormatter instance with the HTML and JSON formatter" do
+        expect(SimpleCov::Formatter::MultiFormatter).to receive(:new).with([
```
I think you could make this a lot easier to test by:

1. passing in the ENV as an argument, that way it's easy to test and we don't need to mess with global state that might leak into other tests for weird results (if the tests here are run in random order so that the nil one isn't run some test further down the line might still have it set)
2. just returning the array of formatters to use instead of the Multiformatter and doing that on the call site, that way you avoid the stubbing :)

fede-moya replied:
Thanks, both suggestions have simplified testing a lot !

PragTob reviewed (commented):
Looking good :+1: 

Small comment about the testing and well feature/acceptance level tests would be great as pointed out in my other comment.

Thanks for implementing all the suggestions so diligently :green_heart: 

![WhatsApp Image 2020-06-07 at 08 16 07a](https://user-images.githubusercontent.com/606517/93805459-121da900-fc48-11ea-836f-30e470931f46.jpeg)

fede-moya commented:
@PragTob thank you very much for your feedback Tobias and for finding time for this !, I'll be trying to wrap this before the end of the current week.

fede-moya commented:
Hey @PragTob 👋🏼  

I have just addressed your latest feedback, added documentation and feature tests. Also I would like to point that you can find more tests for the `JSONFormatter` here https://github.com/codeclimate-community/simplecov_json_formatter/tree/master/spec. 

Let me know If there is anything I should change or add. 

Regards !

efueger commented:
Hi @PragTob 👋 - Hope all is well! Just a quick check-in to see if you need anything else from this side.

PragTob commented:
Hi folks,

no all good. I just had a couple of rough weeks+ in which I didn't have any mind space to dedicate to open source so minus the small bugfix release I didn't touch any OSS in that time so nothing happened here. Sorry about that.

I'll try to cut a release today.

@fede-moya I was aware of the extra tests, but esp. with simplecov being so wide spread I really like to have the additional layer of integration tests to make sure we don't break anything for a whole user run setting. SimpleCov breaks in the most amazing and weird ways.

PragTob commented on features/config_json_formatter.feature:
```diff
@@ -0,0 +1,43 @@
+
+  Background:
+    Given I'm working on the project "faked_project"
+  Scenario: With JSONFormatter
```
separate (comment for myself when fixing it up)

PragTob commented on features/config_json_formatter.feature:
```diff
@@ -0,0 +1,43 @@
+  Scenario: When CC_TEST_REPORTER_ID is set in the environment
+    Given SimpleCov for Test/Unit is configured with:
+      """
+      ENV['CC_TEST_REPORTER_ID'] = "9719ac886877886b7e325d1e828373114f633683e429107d1221d25270baeabf"
```
move into invoking command and make id more generic/show it's fake (comment for myself when fixing it up)

PragTob commented on features/step_definitions/json_steps.rb:
```diff
@@ -0,0 +1,9 @@
+# frozen_string_literal: true
+
+Then /^the JSON coverage report should map:/ do |expected_report|
+  cd(".") do
```
Do we need the `cd`? (comment for myself when fixing it up)

PragTob commented on features/step_definitions/json_steps.rb:
```diff
@@ -0,0 +1,9 @@
+# frozen_string_literal: true
+
+Then /^the JSON coverage report should map:/ do |expected_report|
```
Thanks for putting stuff in a separate file!

PragTob replied:
add `$` to end of rege

PragTob commented on features/step_definitions/json_steps.rb:
```diff
@@ -0,0 +1,9 @@
+
+Then /^the JSON coverage report should map:/ do |expected_report|
+  cd(".") do
+    json_report = File.open("coverage/coverage.json").read
```
File.read? (comment for myself when fixing it up)

PragTob commented on features/step_definitions/json_steps.rb:
```diff
@@ -0,0 +1,9 @@
+  cd(".") do
+    json_report = File.open("coverage/coverage.json").read
+    expected_report = ERB.new(expected_report).result(binding)
+    expect(json_report).to eq(expected_report)
```
potentially make it check against partials, ERB evaluation seems too involved. (comment for myself when fixing it up)

PragTob commented:
Was merged via #950

PragTob commented:
sorry was in a hurry to get this out and felt bad about leaving this here for so long, forgot the customary bunny and thanks :green_heart: 

![ghost](https://user-images.githubusercontent.com/606517/100553704-99fae180-328f-11eb-9c58-799fbf0a2f7a.jpeg)
